transformations/transformations/global_var_offload.py

```diff
@@ -0,0 +1,275 @@
+                if 'update_device' in pragma.content:
+                    if update_device:
+                        pragma_map[pragma] = update_device
+                    else:
+                        pragma_map[pragma] = None
+
+                if 'update_host' in pragma.content:
+                    if update_host:
+                        pragma_map[pragma] = update_host
+                    else:
+                        pragma_map[pragma] = None
+
+        routine.body = Transformer(pragma_map).visit(routine.body)
+
+        # build new imports to add offloaded global vars to driver symbol table
+        new_import_map = {}
+        for s in self._var_set:
+            if s in routine.symbol_map:
+                continue
+
+            if new_import_map.get(self._modules[s], None):
+                new_import_map[self._modules[s]] += as_tuple(s)
+            else:
+                new_import_map.update({self._modules[s]: as_tuple(s)})
+
+        new_imports = ()
+        for k, v in new_import_map.items():
+            new_imports += as_tuple(Import(k, symbols=tuple(Variable(name=s, scope=routine) for s in v)))
+
+        # add new imports to driver subroutine sepcification
+        import_pos = 0
+        if (old_imports := FindNodes(Import).visit(routine.spec)):
+            import_pos = routine.spec.body.index(old_imports[-1]) + 1
+        if new_imports:
+            routine.spec.insert(import_pos, Comment(text=
+               '![Loki::GlobalVarOffload].....Adding global variables to driver symbol table for offload instructions'))
+        import_pos += 1
+        for index, i in enumerate(new_imports):
+            routine.spec.insert(import_pos + index, i)
+
+    def process_kernel(self, routine, **kwargs):
+        """
+        Collect the set of module variables to be offloaded.
+        """
+
+        # filter out procedure imports
+        calls = [c.name for c in FindNodes(CallStatement).visit(routine.body)]
+        calls += FindInlineCalls().visit(routine.body)
+
+        imports = [i for i in FindNodes(Import).visit(routine.spec) if i.symbols not in calls]
+
+        # build map of modules corresponding to imports
+        import_mod = {s.name.lower(): i.module for i in imports for s in i.symbols}
+
+        # separate out derived and basic types
+        basic_types = [s.name.lower() for i in imports for s in i.symbols if s in kwargs['targets']
+                       if isinstance(s.type.dtype, BasicType) and s.name.lower() in self._var_set]
+        deriv_types = [s for i in imports for s in i.symbols if s in kwargs['targets']
+                       if isinstance(s.type.dtype, DerivedType) and s.name.lower() in self._var_set]
+
+        with dataflow_analysis_attached(routine):
+
+            # collect symbols to add to acc update pragmas in driver layer
+            for item in basic_types:
+                if item in routine.body.uses_symbols:
+                    self._acc_copyin.add(item)
```

Comment:
Same comment here: I think this should use `item.trafo_data` to store this analysis, and then merge this in every level of the scheduler traversal from successors. See e.g. the pool allocator transformation for how this works in principle (where it is used to determine the combined size of temporaries under a call hierarchy).

The current implementation would not be optimal with multiple driver entry points within a scheduler traversal as it will always add all variables to the offload directive, even if they aren't used in the relevant call tree.

Reply:
Yes `item.trafo_data` is definitely a better way of doing it, I'll copy the pool allocator pattern of gathering this data from the call tree. I'm implementing this now. Thanks a lot!

Also for some reason I can't reply to your comment about `get_pragma_parameters` so I'll write here: I didn't know that utility existed. Always better to reuse existing code rather than duplicate functionality. I've changed the implementation to use it now, thanks!